File: python/flink_agents/plan/agent_plan.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
#################################################################################
from typing import TYPE_CHECKING, Any, Dict, List, cast

from pydantic import BaseModel, field_serializer, model_validator

from flink_agents.api.agents.agent import Agent
from flink_agents.api.resource import (
    ResourceDescriptor,
    ResourceType,
)
from flink_agents.api.resource_context import ResourceContext
from flink_agents.api.skills import (
    BASH_TOOL,
    LOAD_SKILL_TOOL,
    Skills,
)
from flink_agents.plan.actions.action import Action
from flink_agents.plan.actions.chat_model_action import CHAT_MODEL_ACTION
from flink_agents.plan.actions.context_retrieval_action import CONTEXT_RETRIEVAL_ACTION
from flink_agents.plan.actions.tool_call_action import TOOL_CALL_ACTION
from flink_agents.plan.configuration import AgentConfiguration
from flink_agents.plan.function import PythonFunction
from flink_agents.plan.resource_provider import (
    JavaResourceProvider,
    JavaSerializableResourceProvider,
    PythonResourceProvider,
    PythonSerializableResourceProvider,
    ResourceProvider,
)
from flink_agents.plan.tools.function_tool import from_callable

if TYPE_CHECKING:
    from flink_agents.api.resource import (
        Resource,
    )
    from flink_agents.integrations.mcp.mcp import MCPServer

BUILT_IN_ACTIONS = [CHAT_MODEL_ACTION, TOOL_CALL_ACTION, CONTEXT_RETRIEVAL_ACTION]


class AgentPlan(BaseModel):
    """Agent plan compiled from user defined agent.

    Attributes:
    ----------
    actions: Dict[str, Action]
        Mapping of action names to actions
    actions_by_event : Dict[Type[Event], str]
        Mapping of event types to the list of actions name that listen to them.
    resource_providers: ResourceProvider
        Two level mapping of resource type to resource name to resource provider.
    """

    actions: Dict[str, Action]
    actions_by_event: Dict[str, List[str]]
    resource_providers: Dict[ResourceType, Dict[str, ResourceProvider]] | None = None
    config: AgentConfiguration | None = None

    @field_serializer("resource_providers")
    def __serialize_resource_providers(
        self, providers: Dict[ResourceType, Dict[str, ResourceProvider]]
    ) -> dict:
        # append meta info to help deserialize resource providers
        data = {}
        for type in providers:
            data[type] = {}
            for name, provider in providers[type].items():
                data[type][name] = provider.model_dump()
                if isinstance(provider, PythonResourceProvider):
                    data[type][name]["__resource_provider_type__"] = (
                        "PythonResourceProvider"
                    )
                elif isinstance(provider, PythonSerializableResourceProvider):
                    data[type][name]["__resource_provider_type__"] = (
                        "PythonSerializableResourceProvider"
                    )
                elif isinstance(provider, JavaResourceProvider):
                    data[type][name]["__resource_provider_type__"] = (
                        "JavaResourceProvider"
                    )
                elif isinstance(provider, JavaSerializableResourceProvider):
                    data[type][name]["__resource_provider_type__"] = (
                        "JavaSerializableResourceProvider"
                    )
        return data

    @model_validator(mode="before")
    def __custom_deserialize(self) -> "AgentPlan":
        if "resource_providers" in self:
            providers = self["resource_providers"]
            # restore exec from serialized json.
            if isinstance(providers, dict):
                for type in providers:
                    for name, provider in providers[type].items():
                        if isinstance(provider, dict):
                            provider_type = provider["__resource_provider_type__"]
                            if provider_type == "PythonResourceProvider":
                                self["resource_providers"][type][name] = (
                                    PythonResourceProvider.model_validate(provider)
                                )
                            elif provider_type == "PythonSerializableResourceProvider":
                                self["resource_providers"][type][name] = (
                                    PythonSerializableResourceProvider.model_validate(
                                        provider
                                    )
                                )
                            elif provider_type == "JavaResourceProvider":
                                self["resource_providers"][type][name] = (
                                    JavaResourceProvider.model_validate(provider)
                                )
                            elif provider_type == "JavaSerializableResourceProvider":
                                self["resource_providers"][type][name] = (
                                    JavaSerializableResourceProvider.model_validate(
                                        provider
                                    )
                                )
        return self

    @staticmethod
    def from_agent(agent: Agent, config: AgentConfiguration) -> "AgentPlan":
        """Build a AgentPlan from user defined agent."""
        actions = {}
        actions_by_event = {}
        for action in _get_actions(agent) + BUILT_IN_ACTIONS:
            assert action.name not in actions, f"Duplicate action name: {action.name}"
            actions[action.name] = action
            for event_type in action.listen_event_types:
                if event_type not in actions_by_event:
                    actions_by_event[event_type] = []
                actions_by_event[event_type].append(action.name)

        resource_providers = {}
        for provider in _get_resource_providers(agent, config):
            type = provider.type
            if type not in resource_providers:
                resource_providers[type] = {}
            name = provider.name
            assert name not in resource_providers[type], (
                f"Duplicate resource name: {name}"
            )
            resource_providers[type][name] = provider
        return AgentPlan(
            actions=actions,
            actions_by_event=actions_by_event,
            resource_providers=resource_providers,
            config=config,
        )

    def get_actions(self, event_type: str) -> List[Action]:
        """Get actions that listen to the specified event type.

        Parameters
        ----------
        event_type : Type[Event]
            The event type to query.

        Returns:
        -------
        list[Action]
            List of Actions that will respond to this event type.
        """
        return [self.actions[name] for name in self.actions_by_event[event_type]]

    def get_action_config(self, action_name: str) -> Dict[str, Any]:
        """Get config of the action.

        Parameters
        ----------
        action_name : str
            The name of the action.

        Returns:
        -------
        Dict[str, Any]
            The config of action.
        """
        return self.actions[action_name].config

    def get_action_config_value(self, action_name: str, key: str) -> Any:
        """Get config of the action.

        Parameters
        ----------
        action_name : str
            The name of the action.
        key : str
            The name of the option.

        Returns:
        -------
        Dict[str, Any]
            The option value of the action config.
        """
        return self.actions[action_name].config.get(key, None)


def _get_actions(agent: Agent) -> List[Action]:
    """Extract all registered agent actions from an agent.

    Parameters
    ----------
    agent : Agent
        The agent to be analyzed.

    Returns:
    -------
    List[Action]
        List of Action defined in the agent.
    """
    actions = []
    for name, value in agent.__class__.__dict__.items():
        if isinstance(value, staticmethod) and hasattr(value, "_listen_events"):
            actions.append(
                Action(
                    name=name,
                    exec=PythonFunction.from_callable(value.__func__),
                    listen_event_types=[
                        f"{event_type.__module__}.{event_type.__name__}"
                        for event_type in value._listen_events
                    ],
                )
            )
        elif callable(value) and hasattr(value, "_listen_events"):
            actions.append(
                Action(
                    name=name,
                    exec=PythonFunction.from_callable(value),
                    listen_event_types=[
                        f"{event_type.__module__}.{event_type.__name__}"
                        for event_type in value._listen_events
                    ],
                )
            )
    for name, action in agent.actions.items():
        actions.append(
            Action(
                name=name,
                exec=PythonFunction.from_callable(action[1]),
                listen_event_types=[
                    f"{event_type.__module__}.{event_type.__name__}"
                    for event_type in action[0]
                ],
                config=action[2],
            )
        )
    return actions


def _get_resource_providers(
    agent: Agent, config: AgentConfiguration
) -> List[ResourceProvider]:
    resource_providers = []
    skills_descriptors = {}
    # retrieve resource declared by decorator
    for name, value in agent.__class__.__dict__.items():
        if (
            hasattr(value, "_is_chat_model_setup")
            or hasattr(value, "_is_chat_model_connection")
            or hasattr(value, "_is_embedding_model_setup")
            or hasattr(value, "_is_embedding_model_connection")
            or hasattr(value, "_is_vector_store")
        ):
            if isinstance(value, staticmethod):
                value = value.__func__

            if callable(value):
                descriptor = value()
                if hasattr(descriptor.clazz, "_is_java_resource"):
                    resource_providers.append(
                        JavaResourceProvider.get(name=name, descriptor=value())
                    )
                else:
                    resource_providers.append(
                        PythonResourceProvider.get(name=name, descriptor=value())
                    )

        elif hasattr(value, "_is_tool"):
            if isinstance(value, staticmethod):
                value = value.__func__

            if callable(value):
                # TODO: support other tool type.
                tool = from_callable(func=value)
                resource_providers.append(
                    PythonSerializableResourceProvider.from_resource(
                        name=name, resource=tool
                    )
                )
        elif hasattr(value, "_is_prompt"):
            if isinstance(value, staticmethod):
                value = value.__func__
            prompt = value()
            resource_providers.append(
                PythonSerializableResourceProvider.from_resource(
                    name=name, resource=prompt
                )
            )
        elif hasattr(value, "_is_mcp_server"):
            if isinstance(value, staticmethod):
                value = value.__func__

            descriptor = value()
            _add_mcp_server(name, resource_providers, descriptor, config)
        elif hasattr(value, "_is_skills"):
            if isinstance(value, staticmethod):
                value = value.__func__
            skills_descriptors[name] = value()

    # retrieve resource declared by add interface
    for name, prompt in agent.resources[ResourceType.PROMPT].items():
        resource_providers.append(
            PythonSerializableResourceProvider.from_resource(name=name, resource=prompt)
        )

    for name, tool in agent.resources[ResourceType.TOOL].items():
        resource_providers.append(
            PythonSerializableResourceProvider.from_resource(
                name=name, resource=from_callable(tool.func)
            )
        )

    for name, descriptor in agent.resources[ResourceType.MCP_SERVER].items():
        _add_mcp_server(name, resource_providers, descriptor, config)

    # Merge decorator-based and programmatic skills
    all_skills: Dict[str, Skills] = dict(
        {**skills_descriptors, **agent.resources[ResourceType.SKILLS]}.items()
    )
    _add_skills(all_skills, resource_providers)

    for resource_type in [
        ResourceType.CHAT_MODEL,
        ResourceType.CHAT_MODEL_CONNECTION,
        ResourceType.EMBEDDING_MODEL,
        ResourceType.EMBEDDING_MODEL_CONNECTION,
        ResourceType.VECTOR_STORE,
    ]:
        for name, descriptor in agent.resources[resource_type].items():
            if hasattr(descriptor.clazz, "_is_java_resource"):
                resource_providers.append(
                    JavaResourceProvider.get(name=name, descriptor=descriptor)
                )
            else:
                resource_providers.append(
                    PythonResourceProvider.get(name=name, descriptor=descriptor)
                )

    return resource_providers


def _add_mcp_server(
    name: str,
    resource_providers: List[ResourceProvider],
    descriptor: ResourceDescriptor,
    config: AgentConfiguration,
) -> None:
    provider = PythonResourceProvider.get(name=name, descriptor=descriptor)

    resource_providers.append(provider)

    class ResourceContextPlaceholder(ResourceContext):
        """Placeholder - MCP server construction doesn't need resource resolution."""

        def generate_available_skills_prompt(self, *skill_names: str) -> str:
            pass

        def get_resource(self, name: str, resource_type: "ResourceType") -> "Resource":
            pass

        def get_skill_dirs(self, *skill_names: str) -> List[str]:
            return []

    mcp_server = cast(
        "MCPServer",
        provider.provide(resource_context=ResourceContextPlaceholder(), config=config),
    )

    resource_providers.extend(
        [
            PythonSerializableResourceProvider.from_resource(
                name=prompt.name, resource=prompt
            )
            for prompt in mcp_server.list_prompts()
        ]
    )

    resource_providers.extend(
        [
            PythonSerializableResourceProvider.from_resource(
                name=tool.name, resource=tool
            )
            for tool in mcp_server.list_tools()
        ]
    )

    mcp_server.close()


SKILLS_CONFIG = "_skills_config"


def _add_skills(
    skills_objects: Dict[str, Skills],
    resource_providers: List[ResourceProvider],
) -> None:
    """Register skill configuration and skill tools.

    Merges all Skills objects into a single Skills config resource,
    and registers built-in skill tools (load_skill, bash).


    """
    if len(skills_objects) == 0:
        return

    # Register skill tools via descriptor (no runtime import needed).
    # The tool classes live in flink_agents.runtime.skill_tools and will
    # be instantiated at runtime by PythonResourceProvider.

    resource_providers.extend(
        [
            PythonResourceProvider.get(
                name=LOAD_SKILL_TOOL,
                descriptor=ResourceDescriptor(
                    clazz="flink_agents.runtime.skill.skill_tools.LoadSkillTool",
                ),
            ),
            PythonResourceProvider.get(
                name=BASH_TOOL,
                descriptor=ResourceDescriptor(
                    clazz="flink_agents.plan.tools.bash.bash_tool.BashTool",
                ),
            ),
        ]
    )

    # TODO: Currently, we construct a global agent skill manager for all skill
    #  resource descriptors. In the future, we can support crate individual
    #  agent skill manager for each resource descriptor, and support specifying
    #  skill names and which skill manager they belong to when declaring a chat
    #  model setup. MCP prompts and tools face the same situation, we can refactor
    #  them as a whole.
    paths: List[str] = []
    for skills_obj in skills_objects.values():
        paths.extend(skills_obj.paths)

    merged = Skills.from_local_dir(*dict.fromkeys(paths))

    resource_providers.append(
        PythonSerializableResourceProvider.from_resource(
            name=SKILLS_CONFIG, resource=merged
        )
    )

File: plan/src/main/java/org/apache/flink/agents/plan/AgentPlan.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.agents.plan;

import org.apache.flink.agents.api.Agent;
import org.apache.flink.agents.api.Event;
import org.apache.flink.agents.api.annotation.ChatModelConnection;
import org.apache.flink.agents.api.annotation.ChatModelSetup;
import org.apache.flink.agents.api.annotation.EmbeddingModelConnection;
import org.apache.flink.agents.api.annotation.EmbeddingModelSetup;
import org.apache.flink.agents.api.annotation.Prompt;
import org.apache.flink.agents.api.annotation.Tool;
import org.apache.flink.agents.api.resource.Resource;
import org.apache.flink.agents.api.resource.ResourceDescriptor;
import org.apache.flink.agents.api.resource.ResourceType;
import org.apache.flink.agents.api.resource.SerializableResource;
import org.apache.flink.agents.api.resource.python.PythonResourceAdapter;
import org.apache.flink.agents.api.resource.python.PythonResourceWrapper;
import org.apache.flink.agents.api.tools.ToolMetadata;
import org.apache.flink.agents.plan.actions.Action;
import org.apache.flink.agents.plan.actions.ChatModelAction;
import org.apache.flink.agents.plan.actions.ToolCallAction;
import org.apache.flink.agents.plan.resourceprovider.JavaResourceProvider;
import org.apache.flink.agents.plan.resourceprovider.JavaSerializableResourceProvider;
import org.apache.flink.agents.plan.resourceprovider.PythonResourceProvider;
import org.apache.flink.agents.plan.resourceprovider.ResourceProvider;
import org.apache.flink.agents.plan.serializer.AgentPlanJsonDeserializer;
import org.apache.flink.agents.plan.serializer.AgentPlanJsonSerializer;
import org.apache.flink.agents.plan.tools.FunctionTool;
import org.apache.flink.agents.plan.tools.ToolMetadataFactory;
import org.apache.flink.api.java.tuple.Tuple3;
import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.annotation.JsonSerialize;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;

/** Agent plan compiled from user defined agent. */
@JsonSerialize(using = AgentPlanJsonSerializer.class)
@JsonDeserialize(using = AgentPlanJsonDeserializer.class)
public class AgentPlan implements Serializable {

    /** Mapping from action name to action itself. */
    private Map<String, Action> actions;

    /** Mapping from event class name to list of actions that should be triggered by the event. */
    private Map<String, List<Action>> actionsByEvent;

    /** Two-level mapping of resource type to resource name to resource provider. */
    private Map<ResourceType, Map<String, ResourceProvider>> resourceProviders;

    private AgentConfiguration config;

    private transient PythonResourceAdapter pythonResourceAdapter;

    /** Cache for instantiated resources. */
    private transient Map<ResourceType, Map<String, Resource>> resourceCache;

    public AgentPlan(Map<String, Action> actions, Map<String, List<Action>> actionsByEvent) {
        this.actions = actions;
        this.actionsByEvent = actionsByEvent;
        this.resourceProviders = new HashMap<>();
        this.config = new AgentConfiguration();
        this.resourceCache = new ConcurrentHashMap<>();
    }

    public AgentPlan(
            Map<String, Action> actions,
            Map<String, List<Action>> actionsByEvent,
            Map<ResourceType, Map<String, ResourceProvider>> resourceProviders) {
        this.actions = actions;
        this.actionsByEvent = actionsByEvent;
        this.resourceProviders = resourceProviders;
        this.resourceCache = new ConcurrentHashMap<>();
        this.config = new AgentConfiguration();
    }

    public AgentPlan(
            Map<String, Action> actions,
            Map<String, List<Action>> actionsByEvent,
            Map<ResourceType, Map<String, ResourceProvider>> resourceProviders,
            AgentConfiguration config) {
        this.actions = actions;
        this.actionsByEvent = actionsByEvent;
        this.resourceProviders = resourceProviders;
        this.resourceCache = new ConcurrentHashMap<>();
        this.config = config;
    }

    /**
     * Constructor that creates an AgentPlan from an Agent instance by scanning for all types of
     * annotations.
     *
     * @param agent the agent instance to scan for actions
     * @throws Exception if there's an error creating actions from the agent
     */
    public AgentPlan(Agent agent) throws Exception {
        this(agent, new AgentConfiguration());
    }

    public AgentPlan(Agent agent, AgentConfiguration config) throws Exception {
        this(new HashMap<>(), new HashMap<>());
        extractActionsFromAgent(agent);
        extractResourceProvidersFromAgent(agent);
        this.config = config;
    }

    public void setPythonResourceAdapter(PythonResourceAdapter adapter) {
        this.pythonResourceAdapter = adapter;
    }

    public Map<String, Action> getActions() {
        return actions;
    }

    public Map<String, Object> getActionConfig(String actionName) {
        return actions.get(actionName).getConfig();
    }

    public Object getActionConfigValue(String actionName, String key) {
        return Objects.requireNonNull(actions.get(actionName).getConfig()).get(key);
    }

    public Map<String, List<Action>> getActionsByEvent() {
        return actionsByEvent;
    }

    public Map<ResourceType, Map<String, ResourceProvider>> getResourceProviders() {
        return resourceProviders;
    }

    public List<Action> getActionsTriggeredBy(String eventType) {
        return actionsByEvent.get(eventType);
    }

    /**
     * Get resource from agent plan.
     *
     * @param name the resource name
     * @param type the resource type
     * @return the resource instance
     * @throws Exception if the resource cannot be found or created
     */
    public Resource getResource(String name, ResourceType type) throws Exception {
        // Check cache first
        if (resourceCache.containsKey(type) && resourceCache.get(type).containsKey(name)) {
            return resourceCache.get(type).get(name);
        }

        // Get resource provider
        if (!resourceProviders.containsKey(type)
                || !resourceProviders.get(type).containsKey(name)) {
            throw new IllegalArgumentException("Resource not found: " + name + " of type " + type);
        }

        ResourceProvider provider = resourceProviders.get(type).get(name);

        if (pythonResourceAdapter != null && provider instanceof PythonResourceProvider) {
            ((PythonResourceProvider) provider).setPythonResourceAdapter(pythonResourceAdapter);
        }

        // Create resource using provider
        Resource resource =
                provider.provide(
                        (String anotherName, ResourceType anotherType) -> {
                            try {
                                return this.getResource(anotherName, anotherType);
                            } catch (Exception e) {
                                throw new RuntimeException(e);
                            }
                        });

        // Cache the resource
        resourceCache.computeIfAbsent(type, k -> new ConcurrentHashMap<>()).put(name, resource);

        return resource;
    }

    public AgentConfiguration getConfig() {
        return config;
    }

    public Map<String, Object> getConfigData() {
        return config.getConfData();
    }

    private void writeObject(ObjectOutputStream out) throws IOException {
        String serializedStr = new ObjectMapper().writeValueAsString(this);
        out.writeUTF(serializedStr);
    }

    private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundException {
        String serializedStr = in.readUTF();
        AgentPlan agentPlan = new ObjectMapper().readValue(serializedStr, AgentPlan.class);
        this.actions = agentPlan.getActions();
        this.actionsByEvent = agentPlan.getActionsByEvent();
        this.resourceProviders = agentPlan.getResourceProviders();
        this.config = agentPlan.getConfig();
        this.resourceCache = new ConcurrentHashMap<>();
    }

    private void extractActions(
            Class<? extends Event>[] listenEventTypes, Method method, Map<String, Object> config)
            throws Exception {
        // Convert event types to string names
        List<String> eventTypeNames = new ArrayList<>();
        for (Class<? extends Event> eventType : listenEventTypes) {
            eventTypeNames.add(eventType.getName());
        }

        // Create a JavaFunction for this method
        JavaFunction javaFunction =
                new JavaFunction(
                        method.getDeclaringClass(), method.getName(), method.getParameterTypes());

        // Create an Action
        Action action = new Action(method.getName(), javaFunction, eventTypeNames, config);

        // Add to actions map
        actions.put(action.getName(), action);

        // Add to actionsByEvent map
        for (String eventTypeName : eventTypeNames) {
            actionsByEvent.computeIfAbsent(eventTypeName, k -> new ArrayList<>()).add(action);
        }
    }

    private void addBuiltAction(Action action) {
        // Add to actions map
        actions.put(action.getName(), action);

        // Add to actionsByEvent map
        for (String eventTypeName : action.getListenEventTypes()) {
            actionsByEvent.computeIfAbsent(eventTypeName, k -> new ArrayList<>()).add(action);
        }
    }

    private void extractActionsFromAgent(Agent agent) throws Exception {
        // Add built-in actions
        addBuiltAction(ChatModelAction.getChatModelAction());
        addBuiltAction(ToolCallAction.getToolCallAction());

        // Scan the agent class for methods annotated with @Action
        Class<?> agentClass = agent.getClass();
        for (Method method : agentClass.getDeclaredMethods()) {
            if (method.isAnnotationPresent(org.apache.flink.agents.api.annotation.Action.class)) {
                org.apache.flink.agents.api.annotation.Action actionAnnotation =
                        method.getAnnotation(org.apache.flink.agents.api.annotation.Action.class);

                // Get the event types this action listens to
                Class<? extends Event>[] listenEventTypes =
                        Objects.requireNonNull(actionAnnotation).listenEvents();

                extractActions(listenEventTypes, method, null);
            }
        }

        for (Map.Entry<String, Tuple3<Class<? extends Event>[], Method, Map<String, Object>>>
                action : agent.getActions().entrySet()) {
            Tuple3<Class<? extends Event>[], Method, Map<String, Object>> tuple = action.getValue();
            extractActions(tuple.f0, tuple.f1, tuple.f2);
        }
    }

    private void extractResource(ResourceType type, Method method) throws Exception {
        String name = method.getName();
        ResourceProvider provider;
        ResourceDescriptor descriptor = (ResourceDescriptor) method.invoke(null);
        if (PythonResourceWrapper.class.isAssignableFrom(Class.forName(descriptor.getClazz()))) {
            provider = new PythonResourceProvider(name, type, descriptor);
        } else {
            provider = new JavaResourceProvider(name, type, descriptor);
        }
        addResourceProvider(provider);
    }

    private void extractTool(Method method) throws Exception {
        String name = method.getName();

        // Build parameter type names for reconstruction
        Class<?>[] paramTypes = method.getParameterTypes();

        ToolMetadata metadata = ToolMetadataFactory.fromStaticMethod(method);
        JavaFunction javaFunction =
                new JavaFunction(method.getDeclaringClass(), method.getName(), paramTypes);

        FunctionTool tool = new FunctionTool(metadata, javaFunction);
        JavaSerializableResourceProvider provider =
                JavaSerializableResourceProvider.createResourceProvider(
                        name, ResourceType.TOOL, tool);

        addResourceProvider(provider);
    }

    private void extractResourceProvidersFromAgent(Agent agent) throws Exception {
        Class<?> agentClass = agent.getClass();

        // Scan all fields in the agent class for @Tool and @ChatModel annotations
        for (Field field : agentClass.getDeclaredFields()) {
            field.setAccessible(true); // Allow access to private fields

            String errMsg =
                    "Failed to access field "
                            + field.getName()
                            + " in agent class "
                            + agentClass.getName();

            // Check for @Tool annotation
            if (field.isAnnotationPresent(Tool.class)) {
                String resourceName = field.getName();

                try {
                    Object fieldValue = field.get(agent);
                    if (fieldValue instanceof Resource) {
                        Resource resource = (Resource) fieldValue;
                        ResourceProvider provider =
                                createResourceProvider(
                                        resourceName, ResourceType.TOOL, resource, agentClass);
                        addResourceProvider(provider);
                    }
                } catch (IllegalAccessException e) {
                    throw new Exception(errMsg, e);
                }
            }

            // Check for @ChatModel annotation
            if (field.isAnnotationPresent(ChatModelSetup.class)) {
                ChatModelSetup chatModelAnnotation = field.getAnnotation(ChatModelSetup.class);
                String resourceName = field.getName();

                try {
                    Object fieldValue = field.get(agent);
                    if (fieldValue instanceof Resource) {
                        Resource resource = (Resource) fieldValue;
                        ResourceProvider provider =
                                createResourceProvider(
                                        resourceName,
                                        ResourceType.CHAT_MODEL,
                                        resource,
                                        agentClass);
                        addResourceProvider(provider);
                    }
                } catch (IllegalAccessException e) {
                    throw new Exception(errMsg, e);
                }
            }
        }

        // Scan static methods annotated with @Tool, @Prompt, @ChatModel .etc
        for (Method method : agentClass.getDeclaredMethods()) {
            if (method.isAnnotationPresent(Tool.class)
                    && Modifier.isStatic(method.getModifiers())) {
                extractTool(method);
            } else if (method.isAnnotationPresent(Prompt.class)) {
                String promptName = method.getName();
                SerializableResource prompt = (SerializableResource) method.invoke(null);

                JavaSerializableResourceProvider provider =
                        JavaSerializableResourceProvider.createResourceProvider(
                                promptName, ResourceType.PROMPT, prompt);

                addResourceProvider(provider);
            } else if (method.isAnnotationPresent(ChatModelSetup.class)) {
                extractResource(ResourceType.CHAT_MODEL, method);
            } else if (method.isAnnotationPresent(ChatModelConnection.class)) {
                extractResource(ResourceType.CHAT_MODEL_CONNECTION, method);
            } else if (method.isAnnotationPresent(EmbeddingModelSetup.class)) {
                extractResource(ResourceType.EMBEDDING_MODEL, method);
            } else if (method.isAnnotationPresent(EmbeddingModelConnection.class)) {
                extractResource(ResourceType.EMBEDDING_MODEL_CONNECTION, method);
            }
        }

        for (Map.Entry<ResourceType, Map<String, Object>> entry : agent.getResources().entrySet()) {
            ResourceType type = entry.getKey();
            if (type == ResourceType.CHAT_MODEL || type == ResourceType.CHAT_MODEL_CONNECTION) {
                for (Map.Entry<String, Object> kv : entry.getValue().entrySet()) {
                    ResourceProvider provider;
                    if (PythonResourceWrapper.class.isAssignableFrom(
                            Class.forName(((ResourceDescriptor) kv.getValue()).getClazz()))) {
                        provider =
                                new PythonResourceProvider(
                                        kv.getKey(), type, (ResourceDescriptor) kv.getValue());
                    } else {
                        provider =
                                new JavaResourceProvider(
                                        kv.getKey(), type, (ResourceDescriptor) kv.getValue());
                    }
                    addResourceProvider(provider);
                }
            } else if (type == ResourceType.PROMPT) {
                for (Map.Entry<String, Object> kv : entry.getValue().entrySet()) {
                    JavaSerializableResourceProvider provider =
                            JavaSerializableResourceProvider.createResourceProvider(
                                    kv.getKey(),
                                    ResourceType.PROMPT,
                                    (SerializableResource) kv.getValue());

                    addResourceProvider(provider);
                }
            } else if (type == ResourceType.TOOL) {
                for (Map.Entry<String, Object> kv : entry.getValue().entrySet()) {
                    extractTool(
                            ((org.apache.flink.agents.api.tools.FunctionTool) kv.getValue())
                                    .getMethod());
                }
            }
        }
    }

    /**
     * Creates an appropriate ResourceProvider based on the resource type and whether it's
     * serializable.
     */
    private ResourceProvider createResourceProvider(
            String name, ResourceType type, Resource resource, Class<?> agentClass)
            throws Exception {
        if (resource instanceof SerializableResource) {
            // For serializable resources, use JavaSerializableResourceProvider
            SerializableResource serializableResource = (SerializableResource) resource;
            return JavaSerializableResourceProvider.createResourceProvider(
                    name, type, serializableResource);
        } else {
            throw new UnsupportedOperationException(
                    "Only support declared SerializableResource as field of Agent.");
        }
    }

    /** Adds a resource provider to the resourceProviders map. */
    private void addResourceProvider(ResourceProvider provider) {
        resourceProviders
                .computeIfAbsent(provider.getType(), k -> new HashMap<>())
                .put(provider.getName(), provider);
    }
}
